File: mac_app/Services/Auth/AuthAPIClient.swift
import Foundation

/// HTTP client for authentication API endpoints
final class AuthAPIClient {
    static let shared = AuthAPIClient()

    private let baseURL: URL
    private let session: URLSession
    private let tokenStore = AuthTokenStore.shared

    private init() {
        // Configure base URL based on environment
        let urlString = AppEnvironment.current.apiBaseURL
        self.baseURL = URL(string: urlString)!

        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 60
        self.session = URLSession(configuration: config)
    }

    // MARK: - Device Code Flow

    func requestDeviceCode(deviceInfo: DeviceInfo) async throws -> DeviceCodeResponse {
        let body: [String: Any] = [
            "deviceId": deviceInfo.deviceId,
            "deviceName": deviceInfo.name,
            "platform": deviceInfo.platform
        ]

        return try await post(
            endpoint: "/api/auth/device/code",
            body: body,
            requiresAuth: false
        )
    }

    func pollDeviceCode(_ deviceCode: String) async throws -> DevicePollResponse {
        return try await get(
            endpoint: "/api/auth/device/poll",
            queryItems: [URLQueryItem(name: "deviceCode", value: deviceCode)],
            requiresAuth: false
        )
    }

    // MARK: - Credentials Login

    func login(email: String, password: String, deviceInfo: DeviceInfo) async throws -> LoginResponse {
        let body: [String: Any] = [
            "email": email,
            "password": password,
            "deviceId": deviceInfo.deviceId,
            "deviceName": deviceInfo.name,
            "platform": deviceInfo.platform,
            "osVersion": deviceInfo.osVersion ?? "",
            "appVersion": deviceInfo.appVersion ?? ""
        ]

        return try await post(
            endpoint: "/api/auth/macos/login",
            body: body,
            requiresAuth: false
        )
    }

    // MARK: - Registration

    func register(name: String, email: String, password: String, deviceInfo: DeviceInfo) async throws -> RegistrationResponse {
        let body: [String: Any] = [
            "name": name,
            "email": email,
            "password": password,
            "deviceId": deviceInfo.deviceId,
            "deviceName": deviceInfo.name,
            "platform": deviceInfo.platform,
            "osVersion": deviceInfo.osVersion ?? "",
            "appVersion": deviceInfo.appVersion ?? ""
        ]

        return try await postForRegistration(
            endpoint: "/api/auth/macos/register",
            body: body
        )
    }

    private func postForRegistration(endpoint: String, body: [String: Any]) async throws -> RegistrationResponse {
        let url = baseURL.appendingPathComponent(endpoint)
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw AuthError.serverError("Invalid response")
        }

        // Handle specific registration error codes
        switch httpResponse.statusCode {
        case 200, 201:
            return try JSONDecoder().decode(RegistrationResponse.self, from: data)

        case 400:
            let errorResponse = try? JSONDecoder().decode(RegistrationErrorResponse.self, from: data)
            switch errorResponse?.error {
            case "email_exists":
                throw AuthError.emailAlreadyExists
            case "oauth_account_exists":
                throw AuthError.oauthAccountExists
            case "weak_password":
                throw AuthError.weakPassword(errorResponse?.message ?? "Password does not meet requirements")
            default:
                throw AuthError.serverError(errorResponse?.message ?? "Registration failed")
            }

        default:
            let errorResponse = try? JSONDecoder().decode(RegistrationErrorResponse.self, from: data)
            throw AuthError.serverError(errorResponse?.message ?? "Registration failed")
        }
    }

    // MARK: - Token Refresh

    func refreshTokens(_ refreshToken: String) async throws -> RefreshResponse {
        let body: [String: String] = [
            "refreshToken": refreshToken
        ]

        return try await post(
            endpoint: "/api/auth/macos/refresh",
            body: body,
            requiresAuth: false
        )
    }

    // MARK: - Logout

    func logout(refreshToken: String?, allDevices: Bool = false) async throws {
        let body: [String: Any] = [
            "refreshToken": refreshToken ?? "",
            "allDevices": allDevices
        ]

        let _: EmptyResponse = try await post(
            endpoint: "/api/auth/macos/logout",
            body: body,
            requiresAuth: true
        )
    }

    // MARK: - License Validation

    func validateLicense(deviceId: String) async throws -> License {
        let body: [String: String] = [
            "deviceId": deviceId
        ]

        return try await post(
            endpoint: "/api/license/validate",
            body: body,
            requiresAuth: true
        )
    }

    // MARK: - Usage Recording

    /// Records feature usage to server for server-side tracking
    func recordUsage(action: String, provider: String? = nil, tokensUsed: Int? = nil) async throws {
        var body: [String: Any] = [
            "deviceId": DeviceInfo.current().deviceId,
            "action": action
        ]

        if let provider = provider {
            body["provider"] = provider
        }
        if let tokens = tokensUsed {
            body["tokensUsed"] = tokens
        }

        let _: UsageRecordResponse = try await post(
            endpoint: "/api/usage/record",
            body: body,
            requiresAuth: true
        )
    }

    // MARK: - HTTP Helpers

    private func get<T: Decodable>(
        endpoint: String,
        queryItems: [URLQueryItem] = [],
        requiresAuth: Bool
    ) async throws -> T {
        var components = URLComponents(url: baseURL.appendingPathComponent(endpoint), resolvingAgainstBaseURL: true)!
        if !queryItems.isEmpty {
            components.queryItems = queryItems
        }

        var request = URLRequest(url: components.url!)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if requiresAuth {
            guard let token = await getValidAccessToken() else {
                throw AuthError.notAuthenticated
            }
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        return try await perform(request)
    }

    private func post<T: Decodable>(
        endpoint: String,
        body: [String: Any],
        requiresAuth: Bool
    ) async throws -> T {
        let url = baseURL.appendingPathComponent(endpoint)
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        if requiresAuth {
            guard let token = await getValidAccessToken() else {
                throw AuthError.notAuthenticated
            }
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        return try await perform(request)
    }

    private func perform<T: Decodable>(_ request: URLRequest) async throws -> T {
        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw AuthError.serverError("Invalid response")
        }

        // Handle specific error codes
        switch httpResponse.statusCode {
        case 200...299:
            return try JSONDecoder().decode(T.self, from: data)

        case 401:
            let errorResponse = try? JSONDecoder().decode(ErrorResponse.self, from: data)
            if errorResponse?.error == "oauth_user" {
                throw AuthError.oauthUserNeedsDeviceCode
            }
            throw AuthError.invalidToken

        case 403:
            let errorResponse = try? JSONDecoder().decode(ErrorResponse.self, from: data)
            if errorResponse?.error == "account_blocked" {
                throw AuthError.accountBlocked
            }
            if errorResponse?.error == "device_limit" {
                throw AuthError.deviceLimitReached
            }
            throw AuthError.serverError(errorResponse?.message ?? "Access denied")

        default:
            let errorResponse = try? JSONDecoder().decode(ErrorResponse.self, from: data)
            throw AuthError.serverError(errorResponse?.message ?? "Request failed")
        }
    }

    private func getValidAccessToken() async -> String? {
        // Check if we have a valid access token
        if tokenStore.isAccessTokenValid, let token = tokenStore.accessToken {
            return token
        }

        // Try to refresh
        guard let refreshToken = tokenStore.refreshToken else {
            return nil
        }

        do {
            let response = try await refreshTokens(refreshToken)
            tokenStore.accessToken = response.accessToken
            tokenStore.accessTokenExpiry = Date().addingTimeInterval(TimeInterval(response.expiresIn))
            tokenStore.refreshToken = response.refreshToken
            return response.accessToken
        } catch {
            print("[AuthAPI] Token refresh failed: \(error)")
            return nil
        }
    }
}

// MARK: - Helper Types

private struct ErrorResponse: Decodable {
    let error: String
    let message: String?
}

private struct RegistrationErrorResponse: Decodable {
    let error: String
    let message: String?
    let field: String?
    let requiresDeviceCode: Bool?
}

private struct EmptyResponse: Decodable {}

struct UsageRecordResponse: Decodable {
    let success: Bool
    let recorded: Int
    let usage: UsageStats?
}

struct MagicLinkResponse: Decodable {
    let url: String
    let expiresIn: Int
}

// MARK: - Magic Link

extension AuthAPIClient {
    /// Generate a magic link for auto-login on the web
    func generateMagicLink(redirect: String) async throws -> MagicLinkResponse {
        print("[AuthAPI] Generating magic link for redirect: \(redirect)")
        let body: [String: Any] = ["redirect": redirect]
        do {
            let response: MagicLinkResponse = try await post(
                endpoint: "/api/auth/magic-link/generate",
                body: body,
                requiresAuth: true
            )
            print("[AuthAPI] Magic link generated successfully: \(response.url)")
            return response
        } catch {
            print("[AuthAPI] Magic link generation failed: \(error)")
            throw error
        }
    }
}

// MARK: - Email Check

extension AuthAPIClient {
    /// Check if an email exists and what authentication method it uses
    func checkEmail(_ email: String) async throws -> EmailCheckResponse {
        print("[AuthAPI] Checking email: \(email)")
        let body: [String: Any] = ["email": email]

        return try await postPublic(
            endpoint: "/api/auth/check-email",
            body: body
        )
    }

    /// Post without authentication (for public endpoints)
    private func postPublic<T: Decodable>(
        endpoint: String,
        body: [String: Any]
    ) async throws -> T {
        let url = baseURL.appendingPathComponent(endpoint)
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        return try await perform(request)
    }
}

// MARK: - Google OAuth

extension AuthAPIClient {
    /// Exchange Google authorization code for tokens
    func exchangeGoogleAuth(
        code: String,
        codeVerifier: String,
        redirectUri: String,
        deviceInfo: DeviceInfo
    ) async throws -> GoogleCallbackResponse {
        print("[AuthAPI] Exchanging Google auth code for tokens")

        let body: [String: Any] = [
            "authorizationCode": code,
            "codeVerifier": codeVerifier,
            "redirectUri": redirectUri,
            "deviceId": deviceInfo.deviceId,
            "deviceName": deviceInfo.name,
            "platform": deviceInfo.platform,
            "osVersion": deviceInfo.osVersion ?? "",
            "appVersion": deviceInfo.appVersion ?? ""
        ]

        return try await postGoogleCallback(
            endpoint: "/api/auth/macos/google-callback",
            body: body
        )
    }

    private func postGoogleCallback(
        endpoint: String,
        body: [String: Any]
    ) async throws -> GoogleCallbackResponse {
        let url = baseURL.appendingPathComponent(endpoint)
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw AuthError.serverError("Invalid response")
        }

        switch httpResponse.statusCode {
        case 200, 201:
            let decoded = try JSONDecoder().decode(GoogleCallbackResponse.self, from: data)
            return decoded

        case 400:
            let errorResponse = try? JSONDecoder().decode(GoogleErrorResponse.self, from: data)
            if errorResponse?.error == "credentials_account_exists" {
                throw AuthError.credentialsAccountExists
            }
            throw AuthError.serverError(errorResponse?.message ?? "Google authentication failed")

        case 403:
            let errorResponse = try? JSONDecoder().decode(ErrorResponse.self, from: data)
            if errorResponse?.error == "account_blocked" {
                throw AuthError.accountBlocked
            }
            if errorResponse?.error == "device_limit" {
                throw AuthError.deviceLimitReached
            }
            throw AuthError.serverError(errorResponse?.message ?? "Access denied")

        default:
            let errorResponse = try? JSONDecoder().decode(ErrorResponse.self, from: data)
            throw AuthError.serverError(errorResponse?.message ?? "Google authentication failed")
        }
    }
}

// MARK: - Response Types

struct EmailCheckResponse: Decodable {
    let exists: Bool
    let authMethod: String? // "credentials", "google", or null
}

struct GoogleCallbackResponse: Decodable {
    let accessToken: String
    let refreshToken: String
    let expiresIn: Int
    let user: AuthUser
    let isNewUser: Bool
}

private struct GoogleErrorResponse: Decodable {
    let error: String
    let authMethod: String?
    let message: String?
}

File: win_app_2/src/services/auth/authApiClient.ts
import { getApiBaseUrl } from '../config/appEnvironment'
import { createLogger } from '@/lib/logger'
import type {
  LoginResponse,
  RefreshResponse,
  DeviceCodeResponse,
  DeviceCodePollResponse,
  DeviceInfo,
  EmailCheckResponse,
  License,
} from '@/types/auth'

const log = createLogger('AuthAPI')

class AuthError extends Error {
  constructor(
    message: string,
    public code?: string,
  ) {
    super(message)
    this.name = 'AuthError'
  }
}

export const AUTH_ERRORS = {
  notAuthenticated: new AuthError('Not authenticated', 'not_authenticated'),
  invalidToken: new AuthError('Invalid token', 'invalid_token'),
  tokenExpired: new AuthError('Token expired', 'token_expired'),
  emailAlreadyExists: new AuthError('Email already exists', 'email_exists'),
  oauthAccountExists: new AuthError('Account uses Google Sign-In', 'oauth_account_exists'),
  accountBlocked: new AuthError('Account blocked', 'account_blocked'),
  deviceLimitReached: new AuthError('Device limit reached', 'device_limit'),
}

// Session expiration callback (avoids circular imports with stores)
let onSessionExpiredCallback: (() => void) | null = null

export function setOnSessionExpired(callback: () => void): void {
  onSessionExpiredCallback = callback
}

// In-memory token state
let accessToken: string | null = null
let accessTokenExpiry: number | null = null

export function setAccessToken(token: string, expiresIn: number): void {
  accessToken = token
  accessTokenExpiry = Date.now() + expiresIn * 1000
}

export function clearTokens(): void {
  accessToken = null
  accessTokenExpiry = null
}

function isAccessTokenValid(): boolean {
  // 60-second buffer to avoid edge cases (matches macOS pattern)
  return !!(accessToken && accessTokenExpiry && Date.now() < accessTokenExpiry - 60_000)
}

// Auth-rejection codes mean the refresh token is dead (revoked/expired/blocked).
// The main process already drops the token on these; the renderer just clears
// its in-memory state and surfaces "session expired".
const AUTH_REJECTION_CODES = new Set([
  'invalid_token',
  'token_revoked',
  'token_expired',
  'not_authenticated',
  'device_inactive',
  'account_blocked',
  'device_limit',
  'no_token',
])

async function getValidAccessToken(force = false): Promise<string | null> {
  const cachedValid = isAccessTokenValid()
  const expiryIn = accessTokenExpiry ? accessTokenExpiry - Date.now() : null
  log.info(`[diag] getValidAccessToken force=${force} cachedValid=${cachedValid} expiryInMs=${expiryIn}`)

  if (!force && cachedValid) return accessToken

  // Refresh through the main-process single-flight so concurrent calls (and the
  // 3 windows at boot) coalesce onto one /refresh request and one token write —
  // never spending the one-time-use refresh token more than once at a time.
  log.info('[diag] requesting refresh via main process...')
  const result = await window.electronAPI?.auth.refreshToken(getApiBaseUrl())
  if (!result) {
    log.error('Token refresh failed — no response from main process')
    return null
  }

  if (result.ok) {
    setAccessToken(result.accessToken, result.expiresIn)
    log.info(`[diag] refresh OK — new accessToken len=${result.accessToken.length} expiresIn=${result.expiresIn}s`)
    return result.accessToken
  }

  const isAuthRejection = AUTH_REJECTION_CODES.has(result.code)
  log.warn(`[diag] refresh FAILED — isAuthRejection=${isAuthRejection} code=${result.code}`)

  if (isAuthRejection) {
    log.warn('Token refresh rejected by server, session expired')
    clearTokens()
    onSessionExpiredCallback?.()
  } else {
    // network_error / transient — keep tokens, the next action retries.
    log.error(`Token refresh failed (transient): ${result.code}`)
  }

  return null
}

async function fetchAPI<T>(
  endpoint: string,
  options: {
    method?: string
    body?: Record<string, unknown>
    requiresAuth?: boolean
    queryParams?: Record<string, string>
  } = {},
): Promise<T> {
  const { method = 'GET', body, requiresAuth = false, queryParams } = options
  const baseUrl = getApiBaseUrl()
  let url = `${baseUrl}${endpoint}`

  if (queryParams) {
    const params = new URLSearchParams(queryParams)
    url += `?${params.toString()}`
  }

  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
  }

  if (requiresAuth) {
    const token = await getValidAccessToken()
    if (!token) throw AUTH_ERRORS.notAuthenticated
    headers['Authorization'] = `Bearer ${token}`
  }

  log.info(`API Request: ${method} ${url}`)

  let response: Response
  try {
    // Add timeout to prevent hanging requests
    const controller = new AbortController()
    const timeoutId = setTimeout(() => controller.abort(), 30000) // 30 second timeout

    response = await fetch(url, {
      method,
      headers,
      body: body ? JSON.stringify(body) : undefined,
      signal: controller.signal,
    })

    clearTimeout(timeoutId)
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error)
    log.error(`Network error on ${method} ${url}:`, errorMessage)

    if (errorMessage.includes('aborted')) {
      throw new AuthError('Request timed out. Please check your internet connection.', 'timeout')
    }

    throw new AuthError(
      `Network error: ${errorMessage}. Please check your internet connection.`,
      'network_error'
    )
  }

  log.info(`API Response: ${response.status} ${response.statusText}`)

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}))
    log.error(`API Error ${response.status}:`, errorData)

    if (response.status === 401) {
      if (errorData.error === 'oauth_user') {
        throw new AuthError('This account uses Google Sign-In', 'oauth_user')
      }
      throw AUTH_ERRORS.invalidToken
    }

    if (response.status === 403) {
      if (errorData.error === 'account_blocked') throw AUTH_ERRORS.accountBlocked
      if (errorData.error === 'device_limit') throw AUTH_ERRORS.deviceLimitReached
    }

    if (response.status === 400) {
      if (errorData.error === 'email_exists') throw AUTH_ERRORS.emailAlreadyExists
      if (errorData.error === 'oauth_account_exists') throw AUTH_ERRORS.oauthAccountExists
    }

    throw new AuthError(errorData.message || `Request failed (${response.status})`, errorData.error)
  }

  return response.json()
}

// Device Code Flow
export async function requestDeviceCode(deviceInfo: DeviceInfo): Promise<DeviceCodeResponse> {
  log.info('Requesting device code...')
  const response = await fetchAPI<DeviceCodeResponse>('/api/auth/device/code', {
    method: 'POST',
    body: {
      deviceId: deviceInfo.deviceId,
      deviceName: deviceInfo.deviceName,
      platform: deviceInfo.platform,
    },
  })
  log.info('Device code request successful')
  return response
}

export async function pollDeviceCode(deviceCode: string): Promise<DeviceCodePollResponse> {
  return fetchAPI('/api/auth/device/poll', {
    queryParams: { deviceCode },
  })
}

// Credentials Login
export async function login(
  email: string,
  password: string,
  deviceInfo: DeviceInfo,
): Promise<LoginResponse> {
  return fetchAPI('/api/auth/macos/login', {
    method: 'POST',
    body: {
      email,
      password,
      deviceId: deviceInfo.deviceId,
      deviceName: deviceInfo.deviceName,
      platform: deviceInfo.platform,
      osVersion: deviceInfo.osVersion,
      appVersion: deviceInfo.appVersion,
    },
  })
}

// Registration
export async function register(
  name: string,
  email: string,
  password: string,
  deviceInfo: DeviceInfo,
): Promise<LoginResponse> {
  return fetchAPI('/api/auth/macos/register', {
    method: 'POST',
    body: {
      name,
      email,
      password,
      deviceId: deviceInfo.deviceId,
      deviceName: deviceInfo.deviceName,
      platform: deviceInfo.platform,
      osVersion: deviceInfo.osVersion,
      appVersion: deviceInfo.appVersion,
    },
  })
}

// Token Refresh
export async function refreshTokens(refreshToken: string): Promise<RefreshResponse> {
  return fetchAPI('/api/auth/macos/refresh', {
    method: 'POST',
    body: { refreshToken },
  })
}

// Logout
export async function logout(
  refreshToken: string | null,
  allDevices: boolean = false,
): Promise<void> {
  await fetchAPI('/api/auth/macos/logout', {
    method: 'POST',
    body: { refreshToken: refreshToken || '', allDevices },
    requiresAuth: true,
  })
}

// License Validation
export async function validateLicense(deviceId: string): Promise<License> {
  return fetchAPI('/api/license/validate', {
    method: 'POST',
    body: { deviceId },
    requiresAuth: true,
  })
}

// Usage Recording
export async function recordUsage(action: string, provider?: string): Promise<void> {
  const deviceInfo = await window.electronAPI?.getDeviceInfo()
  await fetchAPI('/api/usage/record', {
    method: 'POST',
    body: {
      deviceId: deviceInfo?.deviceId,
      action,
      ...(provider && { provider }),
    },
    requiresAuth: true,
  })
}

// Email Check
export async function checkEmail(email: string): Promise<EmailCheckResponse> {
  return fetchAPI('/api/auth/check-email', {
    method: 'POST',
    body: { email },
  })
}

// Google OAuth Exchange
export async function exchangeGoogleAuth(
  code: string,
  codeVerifier: string,
  redirectUri: string,
  deviceInfo: DeviceInfo,
): Promise<LoginResponse> {
  return fetchAPI('/api/auth/macos/google-callback', {
    method: 'POST',
    body: {
      authorizationCode: code,
      codeVerifier,
      redirectUri,
      deviceId: deviceInfo.deviceId,
      deviceName: deviceInfo.deviceName,
      platform: deviceInfo.platform,
      osVersion: deviceInfo.osVersion,
      appVersion: deviceInfo.appVersion,
    },
  })
}

export { AuthError, getValidAccessToken, isAccessTokenValid }
